057-run-zk-kraft-clusters-parallel.md

```diff
@@ -0,0 +1,65 @@
+# Allow running ZooKeeper and KRaft based clusters in parallel
+
+This proposal is about changing the way the current `UseKRaft` feature gate works.
+It changes the way the Strimzi Cluster Operator handles ZooKeeper and KRaft based clusters when it is enabled.
+
+## Current situation
+
+Currently, when the `UseKRaft` feature gate is enabled (together with the required `KafkaNodePools` one), the operator expects all the already running Apache Kafka clusters being KRaft-based.
+It means that the `Kafka` custom resources are configured together with `KafkaNodePool`(s) with `broker` and `controller` roles.
+There is no way to differentiate between clusters running in ZooKeeper or KRaft mode.
+When a `Kafka` custom resource is configured to actually use ZooKeeper or it's just badly configured but supposed to be KRaft-based, the operator detects it as having a missing KRaft controllers configuration and logs the following warning:
+
+```shell
+io.strimzi.operator.common.model.InvalidResourceException: The Kafka cluster my-cluster is invalid: [At least one KafkaNodePool with the controller role and at least one replica is required when KRaft mode is enabled]
+```
+
+In this case, the reconciliation fails and the cluster is not operated at all.
+It means that the operator doesn't take any actions on changes applied by the user to the `spec.kafka` or `spec.zookeeper` in the `Kafka` custom resource.
+
+Furthermore, the `UseKRaft` feature gate was created for development and testing.
+As we move forward, it will be soon available for regular clusters and the current state is not going to be sufficient anymore.
+
+## Motivation
+
+Leaving a ZooKeeper-based cluster not operated when the `UseKRaft` feature gate is enabled looks like a bad thing for the user, until the ZooKeeper support will be completely removed.
+Before that, the user should be able to have both ZooKeeper and KRaft based clusters running and operated when the `UseKRaft` feature gate is enabled.
+Also, taking into account the future ZooKeeper to KRaft migration process, as described in the proposal [PR#90](https://github.com/strimzi/proposals/pull/90), it would help the operator migration component to detect which cluster is ZooKeeper-based so "migrate-able" and which one is already KRaft-based.
+
+## Proposal
+
+The proposal is about adding a new `strimzi.io/kraft` annotation to the `Kafka` custom resource, to be applied by the user, in order to identify a ZooKeeper or KRaft based cluster.
+The possible values would be:
+
+* `disabled`, missing or any invalid value: identifies a ZooKeeper-based cluster.
+* `enabled`: identifies a KRaft-based cluster.
+
+This way, during the reconciliation, the operator is able to "detect" a ZooKeeper-based cluster avoiding the warning and allowing the user to operate it.
+On the KRaft side, the annotation would be needed to have the operator reconciling the corresponding `Kafka` custom resource.
```

Comment:
Maybe consider adding a table, to make the behaviour clear with fewer words?

| Operator feature gate | CR Annotation | Operator behaviour |
| --- | --- | --- |
| `-UseKraft` | `enabled` | ??log a warning??|
| `-UseKraft` | anything else | ZK reconciliation | 
| `+UseKraft` | `enabled` | KRaft reconciliation |
| `+UseKraft` | anything else |ZK reconciliation |

Reply:
I can add the table if you like but regarding the first use case `-UseKRaft` with `strimzi.io/kraft: enabled` annotation, I made a comparison with what happens today with node pools (this proposal is aiming to do almost the same but for KRaft feature gate). When we have `-KafkaNodePools` (so no node pools feature gate enabled) but `strimzi.io/node-pools: enabled` on the Kafka CR (deployed together with some `KafkaNodePool`(s)), the operator just ignores that annotation (and node pools) and deploys the Kafka cluster just following the `spec.kafka` and `spec.zookeeper`. No warning printed, it's just ignored, which sounds fine to me imho. @scholzj what are your thoughts on this?

Reply:
I would probably just ignore it when the feature gate is disabled. But not a strong opinion on this one.